### PR TITLE
Improve ranges iteration perf

### PR DESCRIPTION
**Pull request by lofcz**

### CODE REVIEW

File: src/WattleScript.Interpreter/Execution/VM/Processor/Processor_InstructionLoop.cs [hunk 1]
```diff
@@ -1479,25 +1479,32 @@ private void ExecTblInitI(Instruction i)
...
 		}
 
-		private void ExecNewRange()
+		private void ExecNewRange(Instruction i)
 		{
-			DynValue toDv = m_ValueStack.Pop().ToScalar();
-			DynValue fromDv = m_ValueStack.Pop().ToScalar();
+			if (i.NumVal3 == 1) // from v-stack
+			{
+				DynValue toDv = m_ValueStack.Pop().ToScalar();
```
Comment (CallumDev) [inline src/WattleScript.Interpreter/Execution/VM/Processor/Processor_InstructionLoop.cs:1486]: This should just be `m_ValueStack.Pop().CastToInt();` instead of the extraneous ToScalar

File: src/WattleScript.Interpreter/Tree/Expressions/ExprListExpression.cs [hunk 2]
```diff
@@ -6,7 +6,7 @@ namespace WattleScript.Interpreter.Tree.Expressions
 {
 	class ExprListExpression : Expression 
 	{
-		List<Expression> expressions;
+		internal List<Expression> expressions;
```
Comment (CallumDev) [inline src/WattleScript.Interpreter/Tree/Expressions/ExprListExpression.cs:9]: this should change naming to `Expressions` instead of being lowercase

File: src/WattleScript.Interpreter/Tree/Expressions/BinaryOperatorExpression.cs [hunk 3]
```diff
@@ -69,7 +69,7 @@ class LinkedList
 		const Operator NIL_COAL_ASSIGN = Operator.NilCoalescing;
 		const Operator SHIFTS = Operator.BitLShift | Operator.BitRShiftA | Operator.BitRShiftL;
 		const Operator NIL_COAL_INVERSE = Operator.NilCoalescingInverse;
-		const Operator RANGES = Operator.InclusiveRange | Operator.ExclusiveRange | Operator.LeftExclusiveRange | Operator.RightExclusiveRange;
+		internal const Operator RANGES = Operator.InclusiveRange | Operator.ExclusiveRange | Operator.LeftExclusiveRange | Operator.RightExclusiveRange;
```
Comment (CallumDev) [inline src/WattleScript.Interpreter/Tree/Expressions/BinaryOperatorExpression.cs:72]: Instead of exposing all these fields, better to add a method that does the check if it's a range constructor, then return the two expressions. 

`bool TryGetRange(out Expression min, out Expression max, out Operator op)`

File: src/WattleScript.Interpreter/Tree/Statements/ForEachLoopStatement.cs [hunk 4]
```diff
@@ -164,6 +172,66 @@ public override void Compile(FunctionBuilder bc)
 			bc.PopSourceRef();
 		}
 
+		private bool CompilePossibleLiteralRange(Loop l, FunctionBuilder bc)
```
Comment (CallumDev) [inline src/WattleScript.Interpreter/Tree/Statements/ForEachLoopStatement.cs:175]: PossibleLiteral is not quite correct here, as the range itself is not literal expressions. Perhaps `CompileRangeStatement` ?

File: src/WattleScript.Interpreter/Tree/Statements/ForEachLoopStatement.cs [hunk 5]
```diff
@@ -164,6 +172,66 @@ public override void Compile(FunctionBuilder bc)
...
+		{
+			if (!(m_RValues is ExprListExpression listExpr)) return false;
+			if (listExpr.expressions.Count != 1) return false;
+			
+			Expression expr = listExpr.expressions[0];
+			
+			if (!(expr is BinaryOperatorExpression {m_Exp1: { }, m_Exp2: { }} binaryExpr)) return false; // binaryExpr has to have both m_Exp1 and m_Exp2 set
+			if ((binaryExpr.m_Operator & BinaryOperatorExpression.RANGES) == 0) return false;
+			
+			binaryExpr.m_Exp2.Compile(bc); // end
```
Comment (CallumDev) [inline src/WattleScript.Interpreter/Tree/Statements/ForEachLoopStatement.cs:185]: Can check if m_Exp2 is a number literal (TryGetLiteral) and fold in the minus one

File: src/WattleScript.Interpreter/Tree/Statements/ForEachLoopStatement.cs [hunk 6]
```diff
@@ -164,6 +172,66 @@ public override void Compile(FunctionBuilder bc)
...
+
+			if (binaryExpr.m_Operator == Operator.RightExclusiveRange || binaryExpr.m_Operator == Operator.ExclusiveRange) // ..<, >..< -> dec top of stack
+			{
+				bc.Emit_Literal(DynValue.MinusOne);
+				bc.Emit_Operator(OpCode.Add);
+			}
+			
+			new LiteralExpression(lcontext, DynValue.One).Compile(bc); // step
+			
+			binaryExpr.m_Exp1.Compile(bc); // start
```
Comment (CallumDev) [inline src/WattleScript.Interpreter/Tree/Statements/ForEachLoopStatement.cs:195]: Can check if m_Exp1 is a number literal (TryGetLiteral) and fold in the plus one

File: src/WattleScript.Interpreter/Tree/Statements/ForLoopStatement.cs [hunk 7]
```diff
@@ -37,7 +37,7 @@ public ForLoopStatement(ScriptLoadingContext lcontext, Token nameToken, Token fo
 			}
 			else
 			{
-				m_Step = new LiteralExpression(lcontext, DynValue.NewNumber(1));
+				m_Step = new LiteralExpression(lcontext, DynValue.One);
```
Comment (CallumDev) [inline src/WattleScript.Interpreter/Tree/Statements/ForLoopStatement.cs:40]: Remove

File: src/WattleScript.Interpreter/Execution/VM/Processor/Processor_InstructionLoop.cs [hunk 8]
```diff
@@ -1229,7 +1229,7 @@ private int ExecAddStr(int instructionPtr)
 			DataType dvAType = dvA.Type;
 			DataType dvBType = dvB.Type;
 			
-			if (dvA.TryCastToNumber(out var rn) && dvB.TryCastToNumber(out var ln))
+			if (dvA.Type == DataType.Number && dvB.Type == DataType.Number && dvA.TryCastToNumber(out double rn) && dvB.TryCastToNumber(out double ln))
```
Comment (CallumDev) [inline src/WattleScript.Interpreter/Execution/VM/Processor/Processor_InstructionLoop.cs:1232]: See review comment on how to fix this

File: src/WattleScript.Interpreter/DataTypes/DynValue.cs [hunk 9]
```diff
@@ -693,20 +707,17 @@ internal double AssertNumber(int stage)
 		public bool TryCastToNumber(out double d)
 		{
 			ref DynValue rv = ref ScalarReference(ref this);
-			if (rv.Type == DataType.Number)
-			{
-				d = rv.Number;
-				return true;
-			}
-			else if (rv.Type == DataType.String)
+			switch (rv.Type)
```
Comment (CallumDev) [inline src/WattleScript.Interpreter/DataTypes/DynValue.cs:710]: Unnecessary change for this PR, remove.

**Review by CallumDev — CHANGES_REQUESTED** [review]
The additions of the Zero/MinusOne/One fields are not helpful to performance and add extra code currently. Doing a field access is more likely to add a cache miss and penalty from the indirection instead of just constructing the value in place, the .NET JIT is quite efficient with the `NewNumber` code gen. These should be removed from this PR, along with a couple of reformatted methods with no functional changes.

The changes to the AddConcat operator also breaks with tuples (direct type check won't call ToScalar), and then does an extraneous type check afterwards. It would be better to replace that with a method in DynValue like this.

```cs
internal bool TryGetNumber(out double n)
{
    ref DynValue rv = ref ScalarReference(ref this);
    n = rv.m_Number;
    return rv.Type == DataType.Number;
}
```

Finally, randomly exposing the fields in the expressions is quite messy. See review comment for something I would consider cleaner.

File: src/WattleScript.Interpreter/Tree/Statements/ForEachLoopStatement.cs (hunk 4 above)
Comment (lofcz) [reply src/WattleScript.Interpreter/Tree/Statements/ForEachLoopStatement.cs]: Fixed in 8e870ee

File: src/WattleScript.Interpreter/Tree/Statements/ForEachLoopStatement.cs (hunk 5 above)
Comment (lofcz) [reply src/WattleScript.Interpreter/Tree/Statements/ForEachLoopStatement.cs]: Folded in 8e870ee

File: src/WattleScript.Interpreter/Tree/Statements/ForEachLoopStatement.cs (hunk 6 above)
Comment (lofcz) [reply src/WattleScript.Interpreter/Tree/Statements/ForEachLoopStatement.cs]: Folded in 8e870ee

File: src/WattleScript.Interpreter/Tree/Expressions/BinaryOperatorExpression.cs (hunk 3 above)
Comment (lofcz) [reply src/WattleScript.Interpreter/Tree/Expressions/BinaryOperatorExpression.cs]: Did that somehow differently but `internal` is removed. See 8e870ee

File: src/WattleScript.Interpreter/Execution/VM/Processor/Processor_InstructionLoop.cs (hunk 1 above)
Comment (lofcz) [reply src/WattleScript.Interpreter/Execution/VM/Processor/Processor_InstructionLoop.cs]: I need `from/toDv.Type` for errors. ToScalar removed

```
DynValue toDv = m_ValueStack.Pop();
DynValue fromDv = m_ValueStack.Pop();

int? fromInt = fromDv.CastToInt();
int? toInt = toDv.CastToInt();
```

File: src/WattleScript.Interpreter/Tree/Statements/ForLoopStatement.cs (hunk 7 above)
Comment (lofcz) [reply src/WattleScript.Interpreter/Tree/Statements/ForLoopStatement.cs]: Fixed in 4c95935184ba919fceacf7c9de78d2078e4f9abf

File: src/WattleScript.Interpreter/Tree/Expressions/ExprListExpression.cs (hunk 2 above)
Comment (lofcz) [reply src/WattleScript.Interpreter/Tree/Expressions/ExprListExpression.cs]: Fixed in 4c95935184ba919fceacf7c9de78d2078e4f9abf

File: src/WattleScript.Interpreter/Execution/VM/Processor/Processor_InstructionLoop.cs (hunk 8 above)
Comment (lofcz) [reply src/WattleScript.Interpreter/Execution/VM/Processor/Processor_InstructionLoop.cs]: Fixed in aa41897479167bc448ff499436c1ebfef522d654

File: src/WattleScript.Interpreter/DataTypes/DynValue.cs (hunk 9 above)
Comment (lofcz) [reply src/WattleScript.Interpreter/DataTypes/DynValue.cs]: Fixed in a8172ea1ce414989783dccc972befb225d8def58

File: src/WattleScript.Interpreter/Execution/VM/FunctionBuilder.cs [hunk 10]
```diff
@@ -237,6 +237,11 @@ public int Emit_Literal(DynValue value)
 					return AppendInstruction(new Instruction(OpCode.PushString, StringArg(value.String)));
 				case DataType.Range:
 				{
+					if (value.Range.To > Instruction.NumVal2Min && value.Range.To < Instruction.NumVal2Max)
```
Comment (CallumDev) [inline src/WattleScript.Interpreter/Execution/VM/FunctionBuilder.cs:240]: `>=` , `<=` ?

File: src/WattleScript.Interpreter/Tree/Statements/ForEachLoopStatement.cs [hunk 11]
```diff
@@ -164,6 +172,71 @@ public override void Compile(FunctionBuilder bc)
...
+			if (listExpr.Expressions.Count != 1) return false;
+			
+			Expression expr = listExpr.Expressions[0];
+
+			if (!(expr is BinaryOperatorExpression binaryExpr)) return false;
+			if (!binaryExpr.IsRangeCtor()) return false;
+			
+			if (binaryExpr.Operator == Operator.RightExclusiveRange || binaryExpr.Operator == Operator.ExclusiveRange) // ..<, >..< -> sub 1
+			{
+				if (binaryExpr.Exp2 is LiteralExpression le2 && le2.Value.Type == DataType.Number)
```
Comment (CallumDev) [inline src/WattleScript.Interpreter/Tree/Statements/ForEachLoopStatement.cs:187]: The correct way to do things with literal values is to use the `EvalLiteral` function to get the final value. This lets expressions like `(1 + 2)` or `2 ** 4` be passed through as literals. It also makes sense to use TryCastToNumber here to allow for that string coercion.

Finally, when it _isn't_ a literal we do still need those -1 and +1 opcodes to be emitted. Add a test that references variables, arguments to make sure this still holds.

Example:

```cs
if(binaryExpr.Exp2.EvalLiteral(out var le2) && le2.TryCastToNumber(out var leftNum)) {
    bc.Emit_Literal(DynValue.NewNumber((int)(leftNum - 1)));
} else {
    binaryExpr.Exp2.Compile(bc);
    bc.Emit_Literal(DynValue.NewNumber(1));
    bc.Emit_Operator(OpCode.Sub);
}
```

**Review by CallumDev — CHANGES_REQUESTED** [review]
Much better, a couple of small things in implementation

File: src/WattleScript.Interpreter/Execution/VM/FunctionBuilder.cs (hunk 10 above)
Comment (lofcz) [reply src/WattleScript.Interpreter/Execution/VM/FunctionBuilder.cs]: Fixed in 8b1b09eedad6317bdb85325807c3bf721dd271c6

File: src/WattleScript.Interpreter/Tree/Statements/ForEachLoopStatement.cs (hunk 11 above)
Comment (lofcz) [reply src/WattleScript.Interpreter/Tree/Statements/ForEachLoopStatement.cs]: _Should be_ fixed in 8895379822ee7dbcd73e75bfd4dbb369a4f0e929, please take a look whether something else needs to be done. Added a bunch of tests
